src/zcash/NoteEncryption.cpp

```diff
@@ -0,0 +1,169 @@
+#include "NoteEncryption.hpp"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+
+#define NOTEENCRYPTION_CIPHER_KEYSIZE 32
+
+void clamp_curve25519(unsigned char key[crypto_scalarmult_SCALARBYTES])
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &hSig,
+    unsigned char nonce
+   )
+{
+    if (nonce == 0xff) {
+        throw std::logic_error("no additional nonce space for KDF");
+    }
+
+    unsigned char block[128] = {};
+    memcpy(block+0, hSig.begin(), 32);
+    memcpy(block+32, dhsecret.begin(), 32);
+    memcpy(block+64, epk.begin(), 32);
+    memcpy(block+96, pk_enc.begin(), 32);
+
+    unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
+    memcpy(personalization, "ZcashKDF", 8);
+    memcpy(personalization+8, &nonce, 1);
+
+    if (crypto_generichash_blake2b_salt_personal(K, NOTEENCRYPTION_CIPHER_KEYSIZE,
+                                                 block, 128,
+                                                 NULL, 0, // No key.
+                                                 NULL,    // No salt.
+                                                 personalization
+                                                ) != 0)
+    {
+        throw std::logic_error("hash function failure");
+    }
+}
+
+namespace libzcash {
+
+template<size_t MLEN>
+NoteEncryption<MLEN>::NoteEncryption(uint256 hSig) : nonce(0), hSig(hSig) {
+    // All of this code assumes crypto_scalarmult_BYTES is 32
+    // There's no reason that will _ever_ change, but for
+    // completeness purposes, let's check anyway.
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_BYTES);
+    BOOST_STATIC_ASSERT(32 == crypto_scalarmult_SCALARBYTES);
+    BOOST_STATIC_ASSERT(NOTEENCRYPTION_AUTH_BYTES == crypto_aead_chacha20poly1305_ABYTES);
+
+    // Create the ephemeral keypair
+    esk = random_uint256();
```

Comment:
I know I've mentioned this before but I still don't understand why `esk` isn't clamped. If you trace how `sk_enc` is created and used it's a PRF output that gets clamped and then used as the second parameter to `crypto_scalarmult()`. If you trace how `esk` is created and used it's just random and then used as the second parameter to `crypto_scalarmult()` (without ever being clamped). Since they are both generated (roughly) the same way and used the same way I'd expect `esk` to be clamped iff `sk_enc` is.

Am I missing some other use of `sk_enc` which requires it to be clamped?